active/0051-strategies.md

```diff
@@ -0,0 +1,368 @@
++ Start Date: 2017-09-07
++ RFC PR: (leave this empty)
+
+# Summary
+
+The goal of this RFC is to solidify the next version of the mechanism that
+`ember-cli` uses to build final assets. It will allow for a more flexible build
+pipeline for Ember applications. It also unlocks building experimental features
+on top. It is a backward compatible change.
+
+# Motivation
+
+The [Packager
+RFC](https://github.com/chadhietala/rfcs/blob/packager/active/0002-packager.md)
+submitted by [Chad Hietala](https://github.com/chadhietala) is a little over 2
+years old. A lot of things have changed since then and it requires a revision.
+
+The current application build process merges and concatenates input broccoli
+trees. This behaviour is not well documented and is a tribal knowledge. While
+the simplicity of this approach is nice, it doesn't allow for extension. We can
+refactor our build process and provide more flexibility when desired.
+
+Most importantly, the approach described below helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember
+  Engines and Fastboot)
++ unlock a whole different set of plugins we couldn't have before:
+  + ability to create custom bundles (i.e per-engine and per-route bundles)
+  + take advantage of [HTTP2
+    multiplexing](https://http2.github.io/faq/#why-is-http2-multiplexed) and
+    [cache
+    pushing](https://www.mnot.net/blog/2014/01/30/http2_expectations#4-cache-pushing)
+  + optimising plugins (Javascript and CSS treeshaking)
+
+# Scope
+
++ New public API for registering strategies
+
+# Terminology
+
++ **Strategy** - A strategy represents a transformation. Speaking in Broccoli
+  terms, it returns a transformed tree given an input tree.
++ **Assembler** - Responsible for applying strategies (transforms) - default or user
+  provided - to an input tree.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a
+reader to understand.
+
+## Assembler
+
+`Assembler` is similar to a car service shop, where the whole car gets
+"decorated" (repaint, change rims/wheels, turbocharge the engine).
+
+It does not make any assumptions about how the final output should be
+constructed. Instead, it relies on strategies to tell it what the final output
+should look like. This is rather important as it enforces clear separation of
+concerns.
+
+`toTree` method would gather all the trees from the passed in `strategies`, and
+merge them into one final broccoli tree.
+
+## Strategies
+
+Strategy is an extensibility primitive. It gives you granular control over the
```

Comment:
I think it would be nice to provide guidance on how this differs from the similarly existing `app.import` with `using` API.